quick-lint-js: support cross compiling

strager opened the pull request:
quick-lint-js' build system supports cross compiling, but it's not
automatic.

Prior to quick-lint-js version 2.13, cross compiling with Nix works
because some build-time tools are optional. These build-time tools need
a compiler from buildPackages.stdenv. Version 2.13 made these build-time
tools required.

Refactor the quick-lint-js package to use quick-lint-js' cross
compilation support. This fixes cross compilation with Nix.

Upstream documentation for cross compiling:
https://quick-lint-js.com/contribute/build-from-source/cross-compiling/

Testing:

    # Test Linux x86_64 cross compiling for Linux AArch64:
    x86$ uname -a
    Linux strapurp 5.19.0-46-generic #47~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Wed Jun 21 15:35:31 UTC 2 x86_64 x86_64 x86_64 GNU/Linux
    x86$ nix-shell -p pkgsCross.aarch64-multiplatform.quick-lint-js --run 'file $(which quick-lint-js)'
    /nix/store/31npaid1hniy2bb745y9j56r8mm7kmds-quick-lint-js-aarch64-unknown-linux-gnu-2.15.0/bin/quick-lint-js: ELF 64-bit LSB executable, ARM aarch64, version 1 (GNU/Linux), dynamically linked, interpreter /nix/store/f1c3qv0miwni8hx6jm1ym30vfh423nwi-glibc-aarch64-unknown-linux-gnu-2.37-8/lib/ld-linux-aarch64.so.1, for GNU/Linux 3.10.0, not stripped

    # Test the binary on a Linux AArch64 machine:
    x86$ nix --extra-experimental-features nix-command copy --to ssh-ng://parallels@192.168.1.230 /nix/store/31npaid1hniy2bb745y9j56r8mm7kmds-quick-lint-js-aarch64-unknown-linux-gnu-2.15.0/
    arm$ /nix/store/31npaid1hniy2bb745y9j56r8mm7kmds-quick-lint-js-aarch64-unknown-linux-gnu-2.15.0/bin/quick-lint-js --version
    quick-lint-js version 2.15.0

###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

strager committed:
quick-lint-js: support cross compiling

quick-lint-js' build system supports cross compiling, but it's not
automatic.

Prior to quick-lint-js version 2.13, cross compiling with Nix works
because some build-time tools are optional. These build-time tools need
a compiler from buildPackages.stdenv. Version 2.13 made these build-time
tools required.

Refactor the quick-lint-js package to use quick-lint-js' cross
compilation support. This fixes cross compilation with Nix.

Upstream documentation for cross compiling:
https://quick-lint-js.com/contribute/build-from-source/cross-compiling/

Testing:

    # Test Linux x86_64 cross compiling for Linux AArch64:
    x86$ uname -a
    Linux strapurp 6.2.0-35-generic #35~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Fri Oct  6 10:23:26 UTC 2 x86_64 x86_64 x86_64 GNU/Linux
    x86$ nix-shell -p pkgsCross.aarch64-multiplatform.quick-lint-js --run 'file $(which quick-lint-js)'
    /nix/store/3rdxhjxdmiwz1zg2mi477ysc9nb7jsqx-quick-lint-js-aarch64-unknown-linux-gnu-2.17.0/bin/quick-lint-js: ELF 64-bit LSB executable, ARM aarch64, version 1 (GNU/Linux), dynamically linked, interpreter /nix/store/5gdh4mp8rwliq4s33gwcpwzqvsb2xpzr-glibc-aarch64-unknown-linux-gnu-2.38-23/lib/ld-linux-aarch64.so.1, for GNU/Linux 3.10.0, not stripped

    # Test the binary on a Linux AArch64 machine:
    x86$ nix --extra-experimental-features nix-command copy --to ssh-ng://parallels@192.168.4.144 /nix/store/3rdxhjxdmiwz1zg2mi477ysc9nb7jsqx-quick-lint-js-aarch64-unknown-linux-gnu-2.17.0/
    arm$ /nix/store/3rdxhjxdmiwz1zg2mi477ysc9nb7jsqx-quick-lint-js-aarch64-unknown-linux-gnu-2.17.0/bin/quick-lint-js --version
    quick-lint-js version 2.17.0

strager commented:
quick-lint-js version 2.13.0 has released. If we update to 2.13.0 in Nixpkgs without this PR, cross-compiling will be broken. I'd like to get this PR merged before updating quick-lint-js to 2.13.0.

ratsclub reviewed (approved):
It builds just fine here but I have no way of testing cross compilation.

strager commented:
I don't know the Nixpkgs release process. Who do I ask to get this approved PR merged?

Artturin commented on pkgs/development/tools/quick-lint-js/default.nix:
```diff
@@ -29,4 +56,7 @@ stdenv.mkDerivation rec {
     platforms = platforms.all;
   };
+
+  # Expose quick-lint-js-build-tools to nix repl as quick-lint-js.build-tools.
+  build-tools = quick-lint-js-build-tools;
```
should be in `passthru`

Artturin commented on pkgs/development/tools/quick-lint-js/default.nix:
```diff
@@ -12,11 +10,40 @@ stdenv.mkDerivation rec {
+    ];
+    ninjaFlags = "quick-lint-js-build-tools";
+
+    installPhase = ''
```
missing pre and post runHook's

strager replied:
Can you share an example of another package running hooks from installPhase? All the examples I've seen look like this (no extra hooks).

ratsclub replied:
@strager I think he meant something like this:
https://github.com/ratsclub/nixpkgs/blob/8d136b1d4ed0e5f7d661acd171695f9df030b3fd/pkgs/applications/misc/xmrig/default.nix#L54

strager commented:
I updated the patch:
* Rebased onto master.
* Reframed the commit message. Cross compiling worked when I made the patch (quick-lint-js 2.12), but it's broken on master (quick-lint-js 2.15). The commit message used to talk about the upcoming 2.13 release, but now it talks about fixing cross compilation.
* Switched from manual `install` to `cmake --install`. (This was a TODO comment in the original patch.)
* Used `passthru` for `build-tools` per @Artturin's recommendation. 
* Re-ran testing.

strager commented:
I updated the patch again:
* Rebased onto master.
* Called preInstall/postInstall hooks.
* Patch an issue introduced in quick-lint-js 2.16.0 which broke cross-compilation.
* Re-ran testing.

Artturin reviewed (approved):
There's a conflict, doesn't seem that I can push to this branch (no `Maintainers are allowed to edit this pull request.`)

strager commented:
@Artturin Thanks for letting me know. I fixed the conflict and re-tested.
